docs/notebooks/explicit_constraints.pct.py

```diff
@@ -168,27 +145,84 @@ def constraint(input_data):
 # %%
 from trieste.experimental.plotting import plot_bo_points, plot_function_2d
 
-dataset = result.try_get_final_dataset()
-query_points = dataset.query_points.numpy()
-observations = dataset.observations.numpy()
 
-_, ax = plot_function_2d(
-    ScaledBranin.objective,
-    search_space.lower,
-    search_space.upper,
-    grid_density=30,
-    contour=True,
-    figsize=(8, 6),
-)
+def plot_bo_results():
+    dataset = result.try_get_final_dataset()
+    query_points = dataset.query_points.numpy()
+    observations = dataset.observations.numpy()
 
-plot_bo_points(
-    query_points, ax[0, 0], num_initial_points, arg_min_idx, c_pass="green", c_best="purple"
-)
+    _, ax = plot_function_2d(
+        ConstrainedScaledBranin.objective,
+        search_space.lower,
+        search_space.upper,
+        grid_density=30,
+        contour=True,
+        figsize=(8, 6),
+    )
 
-ax[0, 0].contourf(
-    xi, xj, constraint_values, levels=1, colors=[(0.2, 0.2, 0.2, 0.7), (1, 1, 1, 0)], zorder=2
-)
+    plot_bo_points(
+        query_points, ax[0, 0], num_initial_points, arg_min_idx, c_pass="green", c_best="purple"
+    )
 
-ax[0, 0].set_xlabel(r"$x_1$")
-ax[0, 0].set_ylabel(r"$x_2$")
-plt.show()
+    ax[0, 0].contourf(
+        xi, xj, constraint_values, levels=1, colors=[(0.2, 0.2, 0.2, 0.7), (1, 1, 1, 0)], zorder=2
+    )
+
+    ax[0, 0].set_xlabel(r"$x_1$")
+    ax[0, 0].set_ylabel(r"$x_2$")
+    plt.show()
+
+
+plot_bo_results()
+
+# %% [markdown]
+# ## Penalty method
+#
+# ### Acquisition function (penalty method)
+#
+# An alternative to using a constrained optimization method is to construct the _expected constrained improvement_ acquisition function similar to the [inequality-constraints notebook](inequality_constraints.ipynb). However, instead of using probability of feasibility with respect to the constraint model, we construct feasibility from the explicit input constraints. Feasibility is calculated by passing all the constraints residuals (to their respective limits) through a smoothing function and taking the product.
```

Comment:
"through a smoothing function and taking the product."

through a smooth step function and taking the product.